lib/AST/CanonBounds.cpp

```diff
@@ -17,15 +17,127 @@
 #include "llvm/ADT/APInt.h"
 #include "llvm/Support/raw_ostream.h"
 #include "clang/AST/CanonBounds.h"
+#include "clang/AST/ASTContext.h"
 #include "clang/AST/Decl.h"
 #include "clang/AST/Expr.h"
 #include "clang/AST/Stmt.h"
 
 using namespace clang;
 using Result = Lexicographic::Result;
 
-Lexicographic::Lexicographic(ASTContext &Ctx, EqualityRelation *EV) :
-  Context(Ctx), EqualVars(EV), Trace(false) {
+namespace {
+  // Return true if this cast preserve the bits of the value,
+  // false otherwise.
+  static bool IsValuePreserving(CastKind CK) {
+    switch (CK) {
+      case CK_BitCast:
+      case CK_LValueBitCast:
+      case CK_NoOp:
+      case CK_ArrayToPointerDecay:
+      case CK_FunctionToPointerDecay:
+      case CK_NullToPointer:
+      case CK_AssumePtrBounds:
+      case CK_DynamicPtrBounds:
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  static bool isPointerToArrayType(QualType Ty) {
+    if (const PointerType *T = Ty->getAs<PointerType>())
+      return T->getPointeeType()->isArrayType();
+    else
+      return false;
+  }
+  // Ignore operations that don't change runtime values: parens, some cast operations,
+  // array/function address-of and dereference operators, and address-of/dereference
+  // operators that cancel (&* and *&).
+  //
+  // The code for casts is adapted from Expr::IgnoreNoopCasts, which seems like doesn't
+  // do enough filtering (it'll ignore LValueToRValue casts for example).
+  // TODO: reconcile with CheckValuePreservingCast
+  static Expr *IgnoreValuePreservingOperations(ASTContext &Ctx, Expr *E) {
+    while (true) {
```

Comment:
Maybe this could be written in an recursive style? If it works, that's probably best, and there's no reason to re-write it in a recursive style, so there's almost no purpose in leaving this note, except to say that I saw this and appreciated the "get it done" nature of the code but was a little sad at the repeated `continue` paradigm to recur down the `Expr`. I will get over being sad, though.

Reply:
The style was just from the original code that we started from.  I agree that a tail-recursive function would be better.    The continue style is error-prone.   We'll likely extend this code, at which point I'll convert it.